### PR TITLE
Core/Spell: SpellAura Redux

### PR DESCRIPTION
**Pull request by PolarCookie**
**Changes proposed:**
Revisiting https://github.com/TrinityCore/TrinityCore/commit/28288698441524081d583af95f19cb1c9fb90387 and https://github.com/TrinityCore/TrinityCore/commit/a0999744ef02bf1494233411d873552d3d564d8a.
The original issue was that server would crash on multi target delayed spells, because unit auras was being shared between targets and it would cause segmentation faults at times.
Those two commits tried to remedy the problem, but broke spells in different ways.

Solution B would be to delete `Spell::_spellAura` all together, and use `Spell::TargetInfo::HitAura` instead, so far the code is split between their use (notably `SpellScript` uses `Spell::_spellAura`), but they're functionally the same.

**Target branch(es):** 3.3.5

**Issues addressed:** Closes
https://github.com/TrinityCore/TrinityCore/issues/22786

**Tests performed:**
Builds and so far tested
https://www.wowhead.com/spell=33657/resonance
https://www.wowhead.com/spell=48181/haunt
https://www.wowhead.com/spell=59108/glutinous-poison
https://www.wowhead.com/spell=50894/zuldrak-rat
They work, units have their individual stacks and haven't crashed anything yet.

I've not tested Valithria Dreamwalker as reported in https://github.com/TrinityCore/TrinityCore/issues/21609 because I don't know that fight, @Strazhnik could you test this please?

### CODE REVIEW

File: src/server/scripts/Spells/spell_warlock.cpp [hunk 1]
```diff
@@ -773,7 +773,7 @@ class spell_warl_haunt : public SpellScriptLoader
             {
                 if (Aura* aura = GetHitAura())
                     if (AuraEffect* aurEff = aura->GetEffect(EFFECT_1))
-                        aurEff->SetAmount(CalculatePct(aurEff->GetAmount(), GetHitDamage()));
+                        aurEff->SetAmount(CalculatePct(GetHitDamage(), aurEff->GetAmount()));
```
Comment (Warpten) [inline src/server/scripts/Spells/spell_warlock.cpp]: Nice.

Comment (PolarCookie) [reply src/server/scripts/Spells/spell_warlock.cpp]: No it doesn't make a difference, but yes it does matter.

**Comment by jackpoz** [comment]
Could you check with @ariel- if anything is missing too ?

**Comment by kvipka** [comment]
this PR will make a trouble with vehicle spell : 
https://dbwotlk.com/?spell=46598

for example q : [Where in the World is Hemet Nesingwary?](https://dbwotlk.com/?quest=12521) 

will instant entervehicle and exitvehicle

**Comment by PolarCookie** [comment]
@kvipka nice catch!

**Comment by PolarCookie** [comment]
Yes, I have a suspicion might push a commit later today if successful.

**Comment by PolarCookie** [comment]
Update; I've been busy and lazy and haven't put in the time I thought I would, but getting back into it.
What I've found so far; https://db.rising-gods.de/?spell=59319 works.
Note that it's a single effect spell, this will be important.
https://db.rising-gods.de/?spell=46598, https://db.rising-gods.de/?spell=49256 and most other mount spells behave like this:
```
Effect 0: Id 6 (SPELL_EFFECT_APPLY_AURA)
BasePoints = 0
Targets (25, 0) (TARGET_UNIT_TARGET_ANY, NO_TARGET)
Aura Id 236 (SPELL_AURA_CONTROL_VEHICLE), value = 1, misc = 0 (0), miscB = 0, periodic = 0

Effect 1: Id 6 (SPELL_EFFECT_APPLY_AURA)
BasePoints = 0
Targets (1, 0) (TARGET_UNIT_CASTER, NO_TARGET)
Aura Id 4 (SPELL_AURA_DUMMY), value = 1, misc = 0 (0), miscB = 0, periodic = 0
```
Target gets the mount aura, player gets some dummy aura.
The issue comes from https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Unit/Unit.cpp#L3258
All SPELL_AURA_CONTROL_VEHICLE auras are hardcoded to be defined as single target, meaning when player applies his dummy aura the control aura applied to the actual mount is dispelled.
This didn't happen before because of sharing auras between targets.

**Comment by PolarCookie** [comment]
To test, try going to Wyrmrest Temple and do https://db.rising-gods.de/?quest=12372.
Try to mount a the dragon then `.list auras` drake and yourself.

**Comment by Shauren** [comment]
But that is exactly why sharing auras was introduced in the first place - removing CONTROL_VEHICLE from target should also clear the DUMMY from player, the same situation as priest's mind control applying aura to both target and caster

**Comment by PolarCookie** [comment]
Could it be done in other ways? Sharing auras was causing seg faults.

**Comment by PolarCookie** [comment]
I wouldn't call revealing hacks breaking stuff.

**Comment by PolarCookie** [comment]
Entering vehicles work now, I've tested a bunch of them.
Residual dummy auras are not an issue, couldn't find any, because https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Spells/Auras/SpellAuraEffects.cpp#L2976 exists.

Mind controls and possessions work too without any dummy auras left behind, tested most SPELL_AURA_MOD_POSSESS and SPELL_AURA_AOE_CHARM.

In the case that something was missed, it's easy to fix this with spellscripts, not core hacks.

(reserved, testing hunter pet possession)

**Comment by Shauren** [comment]
FYI, you are now reverting 1addad4dbd38d4690c24ad9c07456e06e443c78e

**Comment by PolarCookie** [comment]
Cool, https://github.com/TrinityCore/TrinityCore/issues/9293#issuecomment-19501634 can you reproduce this crash?

**Comment by Shauren** [comment]
I cant remember exactly what you had to do but it might involve that quest in storm peaks that makes you jump between proto drakes

**Comment by PolarCookie** [comment]
Tested The Drakkensryd, could jump between drakes and didn't crash or anything, no residual auras.

**Comment by jackpoz** [comment]
is there a list of spells that one should check when changing this part of the code ? Like a spell that the first PR fixed, another spell that was broken by that PR and that another one fixed etc ? Just to test it once and for all (or until we find another spell that becomes broken)

**Comment by PolarCookie** [comment]
Stuff to test; in no particular order some spells known to have been broken at some point.

**Multi target delayed spells, some with stacking aura**
https://www.wowhead.com/spell=33657/resonance
https://www.wowhead.com/spell=30531/soul-transfer
(```DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 13 AND `SourceEntry` = 30531;``` if you want to test outside of Mag's Lair)
https://www.wowhead.com/spell=59108/glutinous-poison

**Single target stacking aura**
https://www.wowhead.com/spell=50894/zuldrak-rat

https://woehead.way-of-elendil.fr/?spell=63165/decimation
https://www.wowhead.com/spell=48181/haunt

**Assorted quests with either possession or mount**
Don't need to test the quest itself, just if you can mount/possess without leaving any auras on you afterwards, check with `.list auras`
https://www.wowhead.com/quest=9433/a-dip-in-the-moonwell
https://www.wowhead.com/quest=12521/where-in-the-world-is-hemet-nesingwary
https://www.wowhead.com/quest=12983/the-last-of-her-kind
https://www.wowhead.com/quest=12886/the-drakkensryd
https://www.wowhead.com/quest=13086/the-last-line-of-defense
https://www.wowhead.com/quest=11310/warning-some-assembly-required
https://www.wowhead.com/quest=11470/there-exists-no-honor-among-birds
https://www.wowhead.com/quest=12498/on-ruby-wings
https://www.wowhead.com/quest=10612/the-fel-and-the-furious
(consoles are not scripted right, have to manually cast one of the six https://db.rising-gods.de/?spell=38003)
https://www.wowhead.com/quest=12050/lumber-hack
https://www.wowhead.com/item=22728/steam-tonk-controller

Step into Ulduar and try jumping around the vehicle seats there.


**Valithria Dreamwalker**
https://github.com/TrinityCore/TrinityCore/issues/21609

**Comment by PolarCookie** [comment]
Was able to complete Valithria Dreamwalker without crashing, but dunno if it applies to #21609
@ariel- are you still around?

**Comment by PolarCookie** [comment]
https://db.rising-gods.de/?spell=49078 is the vehicle spell for https://db.rising-gods.de/?quest=12326
https://db.rising-gods.de/?spell=47963 is triggered by it and sticks around after leaving the tank.
Found that, and ~40 other vehicle spells I'm going to test a bit further, mostly following the same pattern of a 2nd aura triggered by the vehicle spell which sticks around. 
Busy tonight, but will look into spellscripting it off.

**Comment by jackpoz** [comment]
Nice list, good job :)

**Comment by PolarCookie** [comment]
I've tested almost all spells with `SPELL_AURA_CONTROL_VEHICLE` effect which could have had side effects.
Except Kologarn's Grip (it's WIP in a different PR), and the following list which I couldn't find where were.
ID - 44249 Rope Swing
ID - 55951 Eye Gouge
ID - 60682 Ride Vehicle
ID - 60683 Ride Vehicle
ID - 61423 Mammoth Ride
ID - 61124 Ingest
Most likely they're fine.

Jumping into the Ulduar Salvaged Demolisher catapult causes issue with leftover auras, but that is broken in general and requires it's own PR.

**Comment by jackpoz** [comment]
@PolarCookie is this PR ready to be merged ? are you available to fix some issues if we merge it in the next days ?

**Comment by PolarCookie** [comment]
@jackpoz yes and yes if Netherwing keeps being unstable.

**Comment by PolarCookie** [comment]
Tested Ulduar Vehicles again, specifically Salvaged Siege Engine.
Entering the Siege Engine, logging off and logging on again will leave a residual https://db.rising-gods.de/?spell=65266 on the player. This doesn't happen if you exit the vehicle normally and happens on both this PR and 3.3.5.
https://db.rising-gods.de/?spell=65031 is the aura which links the gear scaling.

The Oculus drakes did not have this issue however
https://db.rising-gods.de/?spell=49464
the player does not receive the linked aura ever for those drakes.
I think the issue is that the player is not supposed to get the 65266 at all, it's for the vehicle.

**Comment by kvipka** [comment]
try to look xt-002:
[code]DoCast(xt002, SPELL_RIDE_VEHICLE_EXPOSED, true);[/code]

ID - 63313 Ride Vehicle

For me it's crash-assert:
ASSERT(!m_vehicle);

**Comment by jackpoz** [comment]
@kvipka could you post the whole crashlog as Gist ? thanks

**Comment by Aokromes** [comment]
maybe unrelated to PR
https://github.com/TrinityCore/TrinityCore/issues/18565

**Comment by kvipka** [comment]
okay
https://gist.github.com/kvipka/ce0c28f42214a23dfaee384da3a12f36

**Comment by kvipka** [comment]
instantly, yes.
Just give few damage at boss and wait while he starting "Tired" Phase and summon Heart.

change : ASSERT(!m_vehicle);
for : if (m_vehicle)

ofc fix it, but it's wrong way =)

**Comment by jackpoz** [comment]
@PolarCookie can you take a look at the crash ? does it happen only with this PR btw ?

**Comment by Keader** [comment]
It will fix https://github.com/TrinityCore/TrinityCore/issues/22920 too.

**Comment by PolarCookie** [comment]
I'll try to take a look at it sometime before Sunday, I did recreate the XT crash but didn't investigate far before getting busy.

**Comment by PolarCookie** [comment]
Based on https://github.com/TrinityCore/TrinityCore/pull/22972#pullrequestreview-194609327 I unmade the spell effect reordering which was the initial plan to reset Spell::_spellAura between targets, later uneeded for this PR by focusing on TargetInfo::HitAura. See the commit log for details.

Now to retest everything, starting with XT

File: src/server/game/Spells/Spell.cpp [hunk 2]
```diff
@@ -3457,7 +3458,7 @@ void Spell::DoProcessTargetContainer(Container& targetContainer)
             if (target.EffectMask & (1 << i))
                 target.DoTargetSpellHit(this, i);
 
-    for (TargetInfoBase& target : targetContainer)
+    for (uint8 i = 0; i < MAX_SPELL_EFFECTS; ++i)
```
Comment (Shauren) [inline src/server/game/Spells/Spell.cpp]: this is definitely not right (does it even build?)

Comment (PolarCookie) [reply src/server/game/Spells/Spell.cpp]: no, working on amending it, forgot to compile first

**Comment by PolarCookie** [comment]
The crash happens at 
https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Entities/Unit/Unit.cpp#L12700
Context is that the XT Heart is trying to switch vehicle seating from slot 0 (the chest) to slot 1 (the exposed point in front of XT's knees). The Heart is "entering" the exposed position while still seated in the chest, that's why `m_vehicle` is not NULL. What should happen is for the heart to first "exit" the chest before "entering" the exposed position.

The reason this doesn't happen in the 3.3.5 branch is because of https://github.com/TrinityCore/TrinityCore/pull/22794/commits/241992b5d890ee7e06932e81360824a5be3f99d0
When `SPELL_AURA_CONTROL_VEHICLE` auras were considered Single Target any of them would be considered single target with any other of them, meaning `SPELL_AURA_CONTROL_VEHICLE` aura A would cancel out `SPELL_AURA_CONTROL_VEHICLE` aura B.
In this particular case https://db.rising-gods.de/?spell=63313 and https://db.rising-gods.de/?spell=63852.

This change was made because most `SPELL_AURA_CONTROL_VEHICLE` spells apply the control aura to their vehicle and a dummy aura to the caster.

Solution might be to expand https://github.com/TrinityCore/TrinityCore/blob/241992b5d890ee7e06932e81360824a5be3f99d0/src/server/game/Entities/Unit/Unit.cpp#L12448-L12456 to handle seat swapping.

**Comment by Killyana** [comment]
It fixes #22051 #22272 also

**Comment by Aokromes** [comment]
imho we can merge it since it looks no one does ulduar looking to the almost 1 year old mimiron unable to be killed and the client freeze xd

**Comment by PolarCookie** [comment]
Not the prettiest, but it works (tm).
Should be reconsidered whenever this happens
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Unit/Unit.cpp#L12564-L12574

Are there any other cases where npc units switch seats on their vehicle?

**Comment by Killyana** [comment]
I just tested this PR and this issue i not fixed https://github.com/TrinityCore/TrinityCore/issues/21810

**Comment by PolarCookie** [comment]
are you sure it's within the scope of this PR? That predates when I started working on the stacking auras.

**Comment by PolarCookie** [comment]
went back to the genesis for me, throwing Zul'Drak Rats still stacks up on the Basilisks.

**Comment by Killyana** [comment]
The issue is about 2 npcs casting the same spell that must stack on player, for example npc=
16901 cast spell=32722

**Comment by PolarCookie** [comment]
I get that, but I don't feel motivated at the time to look into it, can it be given it's own PR?
It doesn't sound like aura sharing which this PR disables, more like aura collision.

**Comment by Killyana** [comment]
Yes it could be related something else.

**Comment by jackpoz** [comment]
@PolarCookie could you merge 335 into this branch ? is there anything else to do on it or is it ready to be merged ? (some things can be delayed to another PR)

**Comment by illfated** [comment]
Rebase this branch on 3.3.5 is just as valid, I presume.

**Comment by PolarCookie** [comment]
retesting tomorrow, have to remake vmaps it seems now.

**Comment by PolarCookie** [comment]
Finished testing stuff, can't find any mistakes from the rebases.
@Killyana this can be merged now; I fixed the bug that I wanted from the beginning and the new crash with XT.

https://github.com/TrinityCore/TrinityCore/pull/22794/files#diff-4522eeffcbc961602e86620011a2d579R12504
should there be a TC_LOG_DEBUG line here same as the exit vehicle some lines earlier?

**Comment by jackpoz** [comment]
Sure, feel free to add some TC_LOG_DEBUG lines here and there. Let me know when it's ready to be merged so I can merge it

**Comment by Killyana** [comment]
There's an issue related to vehicles, when you kill an accessory it's no longer ejected from the vehicle for example: npc 29698 and 27213, riders remains dead on the vehicle once killed, I don't know if it's related to this PR.

**Comment by PolarCookie** [comment]
I'll take a look

**Comment by PolarCookie** [comment]
Happens in 3.3.5, so unrelated.
I'll keep it in mind though.

**Comment by PolarCookie** [comment]
It's happening in 3.3.5 so unrelated.
I'll keep this in mind though.

A thought, this could be the use case for the dummy aura on caster for every vehicle control spell.

**Comment by PolarCookie** [comment]
No, I went through with the debugger; killing the rider triggers the normal dismount path.
It might be related to the same issue that I noticed while fixing "That's Abominable!"; the undead get knocked back and 1 second later die midair, where they suddenly stop moving leaving a host of undead corpses hovering.
There's something about dead units not updating their movement in general?

**Comment by Killyana** [comment]
Yes flying npc remains in the air once dead if they have "CanFly" instead of "DisableGravity", but I don't think it's related as in this case the accessory remains on the vehicle.

**Comment by PolarCookie** [comment]
Doesn't seem to be any error on the vehicle, his passenger guid gets set to null, same with the riders m_vehicle.

**Comment by PolarCookie** [comment]
Here is the issue
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Unit/Unit.cpp#L8674-L8725
At line 8688 ExitVehicle produces a GenericMovementGenerator for the jump off.
Then at 8708-8709 it's replaced by a IdleMovementGenerator

**Comment by PolarCookie** [comment]
No idea what to do with it, so ignoring for this PR.

**Comment by PolarCookie** [comment]
No idea why the dead rider follows with the mount when mount moves, since from the mount's point of view the rider is not there anymore.

**Comment by illfated** [comment]
> No, I went through with the debugger; killing the rider triggers the normal dismount path.
> It might be related to the same issue that I noticed while fixing "That's Abominable!"; the undead get knocked back and 1 second later die midair, where they suddenly stop moving leaving a host of undead corpses hovering.
> There's something about dead units not updating their movement in general?

Thanks for noticing that and mentioning it. I am fairly sure you are right in your assumptions. Something similar can be seen when doing the horde quest(s) https://wow.gamepedia.com/Blinding_the_Eyes_in_the_Sky / https://wow.gamepedia.com/Keeping_the_Alliance_Blind (Daily), at least 1 piece of the "Alliance" airplane stays "dead" in the air at the height where the plane used to be, when the plane visual falls down to the ground.

**Comment by Keader** [comment]
@PolarCookie it's caused by this PR or previous one?
If not... make a ticket (issue) about it...
This PR is ready to merge?

**Comment by PolarCookie** [comment]
@Keader, this PR can be merged, the movement issues is related to something else.
Rider not falling off the mount on death is "fixed" by commenting out 
https://github.com/TrinityCore/TrinityCore/blob/3bc1900018fd38d23e656237e751cbe1ebb8552d/src/server/game/Entities/Unit/Unit.cpp#L8708-L8709
Knockbacked units dying and freezing in mid-air is "fixed" by additionally commenting out
https://github.com/TrinityCore/TrinityCore/blob/3bc1900018fd38d23e656237e751cbe1ebb8552d/src/server/game/Entities/Unit/Unit.cpp#L8711-L8712

**Comment by Killyana** [comment]
There's already 2 issues for that, it's related to moments issues not spells

**Comment by illfated** [comment]
Looks like the PR is complete, then. Movement issues belong to tickets outside of this PR.

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Comment by PolarCookie** [comment]
Thanks for merging!
now do #21772 to complete the circle!

**Pull request merged**